power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/meta_data.hpp

```diff
@@ -51,6 +51,24 @@ template <class T>
 struct ctype_t<T> : ctype_t<std::underlying_type_t<T>> {};
 template <class T> constexpr CType ctype_v = ctype_t<T>::value;
 
+// function selector based on ctype
+// the operator() of the functor should have a single template parameter
+// the selector will instantiate the operator() with relevant type
+template <class Functor, class... Args> decltype(auto) ctype_func_selector(CType ctype, Functor f, Args&&... args) {
+    switch (ctype) {
+    case CType::c_double:
+        return f.template operator()<double>(std::forward<Args>(args)...);
+    case CType::c_double3:
+        return f.template operator()<RealValue<false>>(std::forward<Args>(args)...);
+    case CType::c_int8:
+        return f.template operator()<int8_t>(std::forward<Args>(args)...);
+    case CType::c_int32:
+        return f.template operator()<int32_t>(std::forward<Args>(args)...);
+    default:
+        throw MissingCaseForEnumError{"CType selector", ctype};
+    }
+}
```

Comment:
nice solution!

I know we already use this kind of `switch` logic more in the code. Do we want to change that now as well?

Reply:
The other places are in the validation test. There is a separate ticket to refactor validatin test.